esinstall/src/entrypoints.ts

```diff
@@ -0,0 +1,337 @@
+import {readdirSync, existsSync, realpathSync} from 'fs';
+import path from 'path';
+import validatePackageName from 'validate-npm-package-name';
+import {ExportField, ExportMapEntry, PackageManifestWithExports, PackageManifest} from './types';
+import {parsePackageImportSpecifier, resolveDependencyManifest} from './util';
+import pm from 'picomatch';
+
+// Rarely, a package will ship a broken "browser" package.json entrypoint.
+// Ignore the "browser" entrypoint in those packages.
+const BROKEN_BROWSER_ENTRYPOINT = ['@sheerun/mutationobserver-shim'];
+
+function hasTypes(manifest: PackageManifest): boolean {
+  return !!(manifest.types || manifest.typings);
+}
+
+type FindManifestEntryOptions = {
+  packageLookupFields?: string[];
+  packageName?: string;
+};
+
+/**
+ *
+ */
+export function findManifestEntry(
+  manifest: PackageManifest,
+  entry?: string,
+  {packageLookupFields = [], packageName}: FindManifestEntryOptions = {},
+): string | undefined {
+  let foundEntrypoint: string | undefined;
+
+  if (manifest.exports) {
+    foundEntrypoint =
+      typeof manifest.exports === 'string'
+        ? manifest.exports
+        : findExportMapEntry(manifest.exports['.']);
```

Comment:
I think you’re missing a scenario here (`manifest.exports` is an object but no `['.']` entry):

```json
  "exports": {
    "browser": "./browser.mjs",
    "require": "./index.js"
  }
```

Reply:
I hadn't seen this one but can add a test quickly.

Reply:
Yeah I don’t know of any packages off the top of my head that rely on this, but this is valid according to the Node.js docs: https://nodejs.org/api/packages.html#packages_conditional_exports